core/src/main/scala/kafka/server/DynamicBrokerConfig.scala

```diff
@@ -588,3 +653,106 @@ class DynamicMetricsReporters(brokerId: Int, server: KafkaServer) extends Reconf
+    KafkaConfig.SslTruststorePasswordProp,
+    KafkaConfig.SslKeyManagerAlgorithmProp,
+    KafkaConfig.SslTrustManagerAlgorithmProp,
+    KafkaConfig.SslEndpointIdentificationAlgorithmProp,
+    KafkaConfig.SslSecureRandomImplementationProp,
+    KafkaConfig.SslClientAuthProp,
+
+    // SASL configs
+    KafkaConfig.SaslMechanismInterBrokerProtocolProp,
+    KafkaConfig.SaslJaasConfigProp,
+    KafkaConfig.SaslEnabledMechanismsProp,
+    KafkaConfig.SaslKerberosServiceNameProp,
+    KafkaConfig.SaslKerberosKinitCmdProp,
+    KafkaConfig.SaslKerberosTicketRenewWindowFactorProp,
+    KafkaConfig.SaslKerberosTicketRenewJitterProp,
+    KafkaConfig.SaslKerberosMinTimeBeforeReloginProp,
+    KafkaConfig.SaslKerberosPrincipalToLocalRulesProp
+  )
+}
+
+class DynamicListenerConfig(server: KafkaServer) extends BrokerReconfigurable with Logging {
+
+  override def reconfigurableConfigs: Set[String] = {
+    DynamicListenerConfig.ReconfigurableConfigs
+  }
+
+  def validateReconfiguration(newConfig: KafkaConfig): Unit = {
+
+    def immutableListenerConfigs(kafkaConfig: KafkaConfig, prefix: String): Map[String, AnyRef] = {
+      newConfig.originals.asScala
+        .filterKeys(_.startsWith(prefix))
+        .filterKeys(k => !DynamicSecurityConfigs.contains(k))
+    }
+
+    val oldConfig = server.config
+    val newListeners = listenersToMap(newConfig.listeners)
+    val newAdvertisedListeners = listenersToMap(newConfig.advertisedListeners)
+    val oldListeners = listenersToMap(oldConfig.listeners)
+    val oldAdvertisedListeners = listenersToMap(oldConfig.advertisedListeners)
+    if (!newAdvertisedListeners.keySet.subsetOf(newListeners.keySet))
+      throw new ConfigException(s"Listeners '$newListeners' and advertisedListeners '$newAdvertisedListeners' don't match")
+    if (newListeners.keySet != newConfig.listenerSecurityProtocolMap.keySet)
+      throw new ConfigException(s"Listeners '$newListeners' and listener map '${newConfig.listenerSecurityProtocolMap}' don't match")
+    newListeners.keySet.intersect(oldListeners.keySet).foreach { listenerName =>
+      val prefix = listenerName.configPrefix
+      val newListenerProps = immutableListenerConfigs(newConfig, prefix)
+      val oldListenerProps = immutableListenerConfigs(oldConfig, prefix)
+      if (newListenerProps != oldListenerProps)
+        throw new ConfigException(s"Configs cannot be updated dynamically for existing listener $listenerName, " +
+          "restart broker or create a new listener for update")
+    }
+    if (!newListeners.contains(newConfig.interBrokerListenerName))
+      throw new ConfigException(s"Cannot delete inter-broker listener ${newConfig.interBrokerListenerName}")
+    if (!newAdvertisedListeners.contains(newConfig.interBrokerListenerName))
+      throw new ConfigException(s"Advertised listener must be specified for inter-broker listener ${newConfig.interBrokerListenerName}")
+    else if (!oldAdvertisedListeners.contains(newConfig.interBrokerListenerName))
```

Comment:
If I understand this correctly, we are allowing the inter-broker listener to change, but only to one of the other listeners that are currently advertised. Is that right? Also, the message below could be clarified a little since it doesn't actually mention the inter-broker listener property.

Reply:
I had forgotten that we don't allow changing inter-broker listener at all. I had moved it out to `Future Work` alongside message format version in the KIP. Removed the check.